src/core/quantum_gate.jl

```diff
@@ -93,21 +99,134 @@ q[2]:───────
 ```
 
 """
-abstract type AbstractGate end
+abstract type AbstractGateSymbol end
+
+abstract type AbstractControlledGateSymbol <: AbstractGateSymbol end
 
-abstract type AbstractControlledGate<:AbstractGate end
+get_connected_qubits(gate::AbstractGateSymbol)=[gate.target]
 
-get_connected_qubits(gate::AbstractGate)=[gate.target]
+get_gate_parameters(gate::AbstractGateSymbol)=Dict{String,Real}()
 
-get_gate_parameters(gate::AbstractGate)=Dict{String,Real}()
+get_num_connected_qubits(gate::AbstractGateSymbol)=length(get_connected_qubits(gate))
 
 """
-    is_gate_type(gate::AbstractGate, type::Type)::Bool 
+    Gate
+
+An object representing an `AbstractGateSymbol` inside `QuantumCircuit`.
+
+# Examples
+```jldoctest
+julia> gate_placement = Gate(iswap(1, 2), [1, 2])
+Gate Object: Snowflake.ISwap
+Connected_qubits	: [1, 2]
+Operator:
+(4, 4)-element Snowflake.SwapLikeOperator:
+Underlying data ComplexF64:
+Equivalent DenseOperator:
+1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.0 + 1.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 1.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    1.0 + 0.0im
+```
+
+
+```jldoctest
+julia> gate_placement = Gate(universal(1, pi/2, -pi/2, pi/2), [3])
+Gate Object: Snowflake.Universal
+Parameters:
+theta	: 1.5707963267948966
+phi	: -1.5707963267948966
+lambda	: 1.5707963267948966
+
+Connected_qubits	: [3]
+Operator:
+(2, 2)-element Snowflake.DenseOperator:
+Underlying data ComplexF64:
+0.7071067811865476 + 0.0im    -4.329780281177466e-17 - 0.7071067811865475im
+4.329780281177466e-17 - 0.7071067811865475im    0.7071067811865476 + 0.0im
+```
+"""
+struct Gate <: AbstractGateSymbol
```

Comment:
is the composition pattern here temporary? Seems clumsy to have Gate be a subtype of AbstractGateSymbol, meaning some AbstractGateSymbol have placement information and some don't. Also, if we use a generic struct such as 
```
struct Gate{T} where T<:AbstractGateSymbol
    gate::T
    connected_qubits::Vector{Int}
    (...)
end
```
, this would enable constraining outputs of `compress_to_universal` for instance to `Gate{Universal}`, whereas it now is the all-encompassing `Gate`.

Reply:
Since this is two points

1. Gate should not be inheriting from AbstractGateSymbol
2. Making gates universal makes sense

I am going to answer them separately.

Reply:
1. Yes, it is a stop-gap measure to allow partial implementation of #226. The inheritance will be removed once #226 is complete. I've added a TODO comment to indicate this.

Reply:
2. I agree that it is much cleaner. I am going to start implementing that.

Reply:
2. Gate is now parametric